### PR TITLE
System test to see if caramel is runable

### PR DESCRIPTION
**Pull request by teoti001**
Adding a Makefile that handles the setup and execution of a "complete" system test,
ensuring that caramel can be installed and run on the machine. To conduct the
test it first generate a new CA cert+key using caramel_ca with
tests/ca_test_input.txt, then a new database is initialized with
caramel_initialize_db. After the setup is complete the server is started in the
background with the generated files, the same is then done for caramel_autosign.
After that use client-example.sh to first upload a CSR, then again to check if
the server has processed our CSR, in which case we should have a file
client.csr, if not something has gone wrong. Any failiur to pass stages of the
test results in exitcode 1.

make venv-install : setup a virtual environment(mktemp -d), install caramel here
make systest : run the system test, will install caramel if not already done
make clean : removes the virtual environment(mktemp -d)

Based on #67

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -0,0 +1,75 @@
+#Python3 and virtual environment
+PY3 := $(shell command -v python3 2> /dev/null)
```
Comment (Spindel) [inline Makefile:2]: I don't think this expansion is actually necessary?  If PY3 isn't used as a rule, and rather "assume that python3 is in path" is used, you'll probably have an ok result

File: Makefile [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+VENV := .venv
```
Comment (Spindel) [inline Makefile:3]: This is conceptually unsound. re-using the virtual environment will lead to issues when/if dependencies change, especially if versions of sub-dependencies change.

For a test to be "clean" the venv should be re-initialized each time.

So, it's better to use `VENV := $(mktemp -d)`

File: Makefile [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+PYTHON3 := $(VENV)/bin/python3
+
+#If caramel_tool exists in the venv caramel has been installed
+INSTALLED := $(VENV)/bin/caramel_tool
```
Comment (Spindel) [inline Makefile:7]: Better to give this a name "CARAMEL_TOOL"  instead. That makes reading the rules below out-of-context easier.

File: Makefile [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+		echo "Python 3 could not be found.";\
+		$(RESET_TERM);\
+		exit 2; \
+	fi
+	$(PY3) -m venv $(VENV)
+
+
+#Install the project via setup.py
+.PHONEY: install
+install: $(INSTALLED)
```
Comment (Spindel) [inline Makefile:30]: This rule doesn't actually "make install"  as end-users expect "make install" to work from an traditional way.

prefix it with venv- instead, or remove the "install" alias completely.

File: Makefile [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+		else \
+			$(FAIL); $(LINE); echo "Autosign server exited before time out with exit code: $$?";\
+		fi;
+	@$(BLR)
+
+#Removes the virtual environment created via this makefile
+.PHONEY: clean
+clean:
+	echo "Removing local virtual environment"
+	rm -rf $(VENV)
```
Comment (Spindel) [inline Makefile:75]: Missing newline.

File: Makefile [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+			$(PASS); $(LINE); echo "Autosign server started and timed out successfully";\
+		else \
+			$(FAIL); $(LINE); echo "System test failed: Autosign server exited before time out with exit code: $$?";\
+			exit 1;\
+		fi;
+	@$(BLR)
+
+	@$(BOLD);echo "Start new caramel server and caramel_autosign in the background";$(BLR)
+	setsid $(VENV)/bin/pserve caramel_launcher.ini \
+		ca_cert=$$CARAMEL_CA_CERT \
```
Comment (Spindel) [inline Makefile:83]: Why are these necessary? Don't the tools take the CARAMEL_CA_CERT environment variable?

File: Makefile [hunk 7]
```diff
@@ -0,0 +1,123 @@
...
+	sleep 2s #Give server some time to start
+	@$(BLR)
+
+	@ $(BOLD); echo "Use client-example.sh to first upload our CSR, wait for it \
+	to get processed,then call it again to confirm the server stored our CSR"; $(BLR)
+	chmod +x client-example.sh
+	./client-example.sh $(VENV)
+	sleep 2s
+	./client-example.sh $(VENV)
+	kill `cat $(VENV)/server.pid` `cat $(VENV)/autosign.pid`
```
Comment (Spindel) [inline Makefile:105]: Mixing shell syntax and make syntax here. $(shell   )  is usually recommended for shell syntax. vs backticks.


Also. Since you have PID files after launching, you can make the launch server a make target that creates the PID file, which in turn willl clean out the dep chain here.

File: Makefile [hunk 8]
```diff
@@ -0,0 +1,123 @@
...
+	@$(BOLD); echo "Removing $(CARAMEL_CA_CERT), $(CARAMEL_CA_KEY) and $(CARAMEL_DBURL)"; $(LINE); $(RESET_TERM)
+	rm -f $(CARAMEL_CA_CERT) $(CARAMEL_CA_KEY) $(CARAMEL_DBURL)
+	@$(BLR)
+
+	@$(BOLD); echo "Generate new CA cert and key with ca_test_input.txt"; $(LINE); $(RESET_TERM)
+	$(VENV)/bin/caramel_ca < ca_test_input.txt
+	@$(BLR)
+
+	@$(BOLD); echo "Create a new DB at $(CARAMEL_DBURL)"; $(LINE); $(RESET_TERM);
+	$(VENV)/bin/caramel_initialize_db
```
Comment (Spindel) [inline Makefile:52]: So. This command actually creates a file.  That makes it an excellent case for it's own target.


```make
$(DB_FILE): $(VENV)
   $(VENV)/bin/caramel_initialize_db
``` 

And then the server can be:

```make
$(SERVER_PID): $(DB_FILE) $(VENV)
     $(VENV)/bin/pserve caramel_launcher.ini
```

File: client-example.sh [hunk 9]
```diff
@@ -76,6 +75,7 @@ fi
 
 if [ $STATUS -eq 404 ];
 then
-    rm -f $CSRSUM
-    curl ${CURL_OPTS} --data-binary @$MY_CSR ${POST_URL}/${CSRSUM}
+  rm -f $1/$CSRSUM
```
Comment (Spindel) [inline client-example.sh:78]: inconsistent quoting with  above statements.  

Also, never a good idea to refer to $1 more than once in a single script. If it's worth using, give it a name.

SUBDIR="$1"  or so.

**Review by Spindel — COMMENTED** [review]
Test looks okay. Some issues with adding more stuff to the root of the project rather than in a subdirectory for tests.

File: Makefile (hunk 6 above)
Comment (teoti001) [reply Makefile:83]: This is not a tool, pserve does not look for our environment variables. But it should probably use the launcher instead, which reads our variables.

File: client-example.sh [hunk 10]
```diff
@@ -60,11 +60,11 @@ fi
 #  if 404
 #       upload CSR
 
-STATUS=$(curl ${CURL_OPTS}  -w '%{http_code}' --url ${POST_URL}/${CSRSUM} -o ${CSRSUM})
+STATUS=$(curl ${CURL_OPTS} -w '%{http_code}' --url ${POST_URL}/"${CSRSUM}" -o "${1}/${CSRSUM}")
```
Comment (Spindel) [inline client-example.sh:63]: One $1 left here.

File: client-example.sh [hunk 11]
```diff
@@ -76,6 +76,7 @@ fi
 
 if [ $STATUS -eq 404 ];
 then
-    rm -f $CSRSUM
-    curl ${CURL_OPTS} --data-binary @$MY_CSR ${POST_URL}/${CSRSUM}
+  rm -f $CA_DIR/$CSRSUM
```
Comment (Spindel) [inline client-example.sh:79]: Quoting missing here.  Space , *  or other fun chars in $1

File: caramel_launcher.sh [hunk 12]
```diff
@@ -0,0 +1,12 @@
+#! /bin/sh
```
Comment (Spindel) [inline caramel_launcher.sh:1]: Move the launcher and ini to tests/  as well. pserve isn't production-recommended as it defaults to waitress as the server.

**Comment by Spindel** [comment]
LGTM.
